docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,342 @@
+
+Regarding ClusterClass, this option functions the same as [Option 2] or [Option 3], where all ASO fields can
+be defined in a template. This option opens up an additional safeguard though, where webhooks could flag
+fields which should not be defined in a template. Similar webhook checks would be less practical in the other
+options where CAPZ would need to be aware of the set of disallowed fields for each ASO API version that a user
+could use.
+
+#### Decision
+
+We are opting to move forward with [Option 4]. That approach is most consistent with how the API is defined
+today and would make for the smoothest transition for users.
+
+### Security Model
+
+Overall, none of the approaches outlined above change what data is ultimately represented in the API, only the
+higher-level shape of the API. That means there is no further transport or handling of secrets or other
+sensitive information beyond what CAPZ already does.
+
+### Risks and Mitigations
+
+Increasing CAPZ's reliance on ASO and exposing ASO to users at the API level further increases the risk
+[previously discussed](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/ce3c130266b23a8b67aa5ef9a21f257ff9e6d63e/docs/proposals/20230123-azure-service-operator.md?plain=1#L169)
+that since ASO has not yet been proven to be as much of a staple as the other projects that manage
+infrastructure on Azure, ASO's lifespan may be more limited than others. If ASO were to sunset while CAPZ
+still relies on it, CAPZ would have to rework its APIs. This risk is mitigated by the fact that no
+announcements have yet been made regarding ASO's end-of-life and the project continues to be very active.
+
+## Alternatives
+
+The main alternative to leveraging ASO to expose the entire AKS API from CAPZ would be for CAPZ to build its
+own code generation pipeline which takes as input the Azure API specs and ultimately produces definitions for
+AzureManagedControlPlane and AzureManagedMachinePool and mappings to ASO's ManagedCluster and
+ManagedClustersAgentPool. Such a solution would likely end up being too similar to what ASO already provides
+for the effort to be worthwhile.
+
+## Upgrade Strategy
+
+Each of the [options above](#api-design-options) suggest additive, backwards-compatible API changes. Users
+will not need to perform any action after upgrading to the first version of CAPZ that implements this
+proposal.
+
+The new API field allowing the entire ASO resource to be defined in the CAPZ API will overlap with several
+existing CAPZ API fields. e.g. with [Option 4], AzureManagedControlPlane's `spec.location` maps to the same
+ASO API field as `spec.managedClusterSpec.location`. If both are defined, then the `spec.managedCluster.*`
+field takes precedence. If only `spec.location` is defined, its value will still be used to construct the
+desired state for the ManagedCluster.
+
+An alternative available with [Option 3] and [Option 4] is to introduce a new CAPZ API version for
```

Comment:
One possible downside of this from a UX perspective is that ManagedCluster is _big_ and getting bigger all the time. If users _must_ delve into `ASOManagedClusterSpec` they could possibly be overwhelmed by how big it is... OTOH with good samples/examples possibly not a problem.

Reply:
> One possible downside of this 

Not sure I'm totally following. What exactly is "this" here?

> If users _must_ delve into `ASOManagedClusterSpec` they could possibly be overwhelmed by how big it is... 

It seems like this is the case with any of the options, right? Since they all expose the entire ASO spec.

Reply:
"This" was moving the existing CAPZ properties down into this new structure.

Reply:
Ah, gotcha. I think as long as we have conversion webhooks set up in this case, then users wouldn't be forced into that change any time soon. I don't think we've gotten too creative in how we structure the CAPZ types, so hopefully transitioning when the time comes isn't too much of a burden.